### PR TITLE
[SYCL] Return trivial default constructor to half type

### PR DESCRIPTION
**Pull request by maximdimakov**
Return trivial default constructor to half type. Without default constructor it is impossible to create union with half type. 
Changed implementation of marray and half_vec constructors for optimizing them to use half with trivial default constructor.
Update files accordingly to returned half trivial default constructor.
This patch corrects https://github.com/intel/llvm/pull/4518

Signed-off-by: mdimakov <maxim.dimakov@intel.com>

### CODE REVIEW

**Review by vladimirlaz — CHANGES_REQUESTED** [review]
This difference was introduced intentionally to resolve mangling issues with different compilers. Please do not merge until resolving all mangling concerns

**Review by vladimirlaz — DISMISSED** [review]

**Review by vladimirlaz — DISMISSED** [review]

**Comment by maximdimakov** [comment]
@AlexeySachkov could you, please, review?

File: sycl/include/CL/sycl/half_type.hpp [hunk 1]
```diff
@@ -40,7 +40,6 @@ namespace detail {
 // bitselect builtin uses union with half specialization, but half has no
 // trivial constructor. This struct allow to use host half class instead of main
 // half class in the bitselect builtin.
```
Comment (AlexeySachkov) [inline sycl/include/CL/sycl/half_type.hpp:42]: I guess this comment should be removed as well

File: sycl/test/abi/sycl_symbols_windows.dump [hunk 2]
```diff
@@ -341,7 +341,6 @@
 ??0half_v2@host_half_impl@detail@sycl@cl@@QEAA@AEBM@Z
 ??0half_v2@host_half_impl@detail@sycl@cl@@QEAA@AEBV01234@@Z
 ??0half_v2@host_half_impl@detail@sycl@cl@@QEAA@G@Z
-??0half_v2@host_half_impl@detail@sycl@cl@@QEAA@XZ
```
Comment (AlexeySachkov) [inline sycl/test/abi/sycl_symbols_windows.dump:344]: @alexbatashev, is this considered to be an ABI break?

Comment (maximdimakov) [reply sycl/test/abi/sycl_symbols_windows.dump:344]: This line was introduced in https://github.com/intel/llvm/pull/4518 . It is not in the latest release

File: sycl/include/CL/sycl/half_type.hpp (hunk 1 above)
Comment (maximdimakov) [reply sycl/include/CL/sycl/half_type.hpp]: Removed

**Review by AlexeySachkov — APPROVED** [review]

**Review by vladimirlaz — APPROVED** [review]

**Pull request merged**